### PR TITLE
Added hash prefix implementation (server code not yet merged) 

### PR DESCRIPTION
**Pull request by Joe-Dowd**
This feature changes the way segments are obtained from the server in a way that improves user privacy, but disables fetching sgements while the video is being played should they be submitted. 

Instead of requesting segments using the video ID (the string after watch?v=), the video ID is hashed and the first 4 characters of the hash are sent to the server, which returns all videos where the hash also starts with those characters. The client then selects the video it wants (if any) and uses the segments provided in that request.

### CODE REVIEW

**Comment by Joe-Dowd** [comment]
https://github.com/ajayyy/SponsorBlockServer/pull/127

File: src/content.ts [hunk 1]
```diff
@@ -618,12 +618,36 @@ function sponsorsLookup(id: string) {
...
+    } else {
+        getRequest = utils.asyncRequestToServer('GET', "/api/skipSegments", {
+            videoID: id,
+            categories
+        });
+    }
+    getRequest.then(async (response: FetchResponse) => {
         if (response?.ok) {
-            let recievedSegments: SponsorTime[] = JSON.parse(response.responseText);
+            let getResult = JSON.parse(response.responseText);
```
Comment (ajayyy) [inline src/content.ts:635]: Maybe the variable from `getResult` to `result`?

File: src/content.ts [hunk 2]
```diff
@@ -618,12 +618,36 @@ function sponsorsLookup(id: string) {
...
+            if (Config.config.hashPrefix) {
+                getResult = getResult.filter((video) => {
+                    return video.videoID = id;
+                });
+                if (getResult.length === 1) {
```
Comment (ajayyy) [inline src/content.ts:640]:
```suggestion
                if (getResult.length > 0) {
```

File: src/utils.ts [hunk 3]
```diff
@@ -378,6 +379,20 @@ class Utils {
...
+
+        for (let i = 0; i < times; i++) {
+            let hash = sha256.create();
+            hash.update(value);
+            hash.hex();
+            value = hash.toString();
+        }
+
+        return value;
+    }
```
Comment (ajayyy) [inline src/utils.ts:394]: This is great to add, as we'll be able to pre-calculate the public userID on the frontend now.

**Review by ajayyy — CHANGES_REQUESTED** [review]

**Comment by ajayyy** [comment]
On my local environment, segments are loading very slowly (Server-side).

**Comment by ajayyy** [comment]
Index fixed it

File: src/utils.ts [hunk 4]
```diff
@@ -378,6 +379,20 @@ class Utils {
     isFirefox(): boolean {
         return typeof(browser) !== "undefined";
     }
+
+    getHash(value: string, times=5000): string {
+        if (times <= 0) return "";
+
+        for (let i = 0; i < times; i++) {
```
Comment (xPaw) [inline src/utils.ts:386]: What is the idea behind hashing it 5k times?

And take a look at native function: https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest

Comment (Joe-Dowd) [reply src/utils.ts:386]: That's a default value taken from the server code 
https://github.com/ajayyy/SponsorBlockServer/blob/master/src/utils/getHash.js#L3-L12

On the server side, userIDs are hashed 5000 times before being stored in the database.

Comment (ajayyy) [reply src/utils.ts:386]: It is to be the same as the server, which hashes 5k times for it's operation.

That's a good idea to use the browser crypto function, thanks!

Comment (xPaw) [reply src/utils.ts:386]: But what is the reason for hashing it more than once?

Comment (ajayyy) [reply src/utils.ts:386]: It was to make brute forcing harder since the hash of the private userID is in the public database.

Comment (xPaw) [reply src/utils.ts:386]: Fair enough, I didn't spot that it's called with times=1 for video IDs.

**Pull request merged**